Details and Footer components

stephaniehobson opened the pull request:
- Creates `Mzp.Utils` and `Mzp.Supports` objects
- Add mixin to transform svgs into data uris
  - And move expand icons to use mixin
- Add details component
  - can be initialized on headings or as a `<details>` polyfill
  - including demo page
- Default styling for details & summary elements
- Add footer component
  - uses details component to expand and collapse menu on mobile

Closes #184 Closes #129

stephaniehobson commented on src/assets/js/protocol/protocol-details.js:
```diff
@@ -0,0 +1,229 @@
+        if (fake) {
+            root.parentNode.removeChild(root);
+        }
+        return diff; //todo, how do I document return values
```
How do I document functions with return values?

alexgibson replied:
You can to `@return`, like here https://github.com/mozilla/protocol/blob/master/src/assets/js/protocol/protocol-lang-switcher.js#L20

stephaniehobson commented on src/assets/js/protocol/protocol-details.js:
```diff
@@ -0,0 +1,229 @@
+            summary.parentNode.insertBefore(details, summary.nextSibling);
+        } else {
+            // error
+            // TODO: what do we do with errors? fail silently?
```
What do we do with errors? Fail silently?

alexgibson replied:
I guess It depends how bad an error could be. Throwing an error is useful for developers, but it could also halt execution of JS which could be even worse than the original error. Maybe the best thing to do is to try and fail gracefully here?

alexgibson replied:
I guess It depends how bad an error could be. Throwing an error is useful for developers, but it could also halt execution of JS which could be even worse than the original error. Maybe the best thing to do is to try and fail gracefully here?

stephaniehobson replied:
I moved stuff around so I could halt initialization if something goes wrong.

stephaniehobson commented on src/assets/js/protocol/protocol-details.js:
```diff
@@ -0,0 +1,229 @@
+        }
+
+        // close by default
+        // TODO: add support for open attribute
```
This isn't necessary for the footer to launch, can I leave this todo?

stephaniehobson commented on src/assets/js/protocol/protocol-details.js:
```diff
@@ -0,0 +1,229 @@
+    if(!Details.isSupported) {
+        // not supported, add support
+        Details.init('summary');
+        // TODO are some aria things necissary with a passing feature test? tabindex? role=button?
```
Note for myself to do some research

stephaniehobson replied:
Looks like support is all or nothing at this point!

stephaniehobson commented on src/assets/js/protocol/protocol-utils.js:
```diff
@@ -0,0 +1,59 @@
+
+    // matches() polyfill
+    // TODO: where should this go?
+    if (!Element.prototype.matches) {
```
Is there a better place to put a polyfill?

alexgibson replied:
Hmm, we don't yet use any polyfills in Protocol, so we don't have a place to store them. We should also document what polyfills we're using which will be really important (maybe something under documentation in the menu?).

This particular piece of code doesn't seem to be a polyfill though, it's more smushing vendor prefixes into a single reference. There does seem to be a [polyfill listed on MDN](https://developer.mozilla.org/en-US/docs/Web/API/Element/matches#Polyfill) though?

stephaniehobson replied:
Maybe it's more of a vendorfill? Is that a thing?

stephaniehobson replied:
Leaving this here for now in the interests of getting the footer into use.

stephaniehobson commented on src/assets/js/protocol/protocol-utils.js:
```diff
@@ -0,0 +1,59 @@
+     * @param {String} filter - Selector to filter results if you only want siblings matching this pattern
+     * - https://gomakethings.com/how-to-get-all-sibling-elements-until-a-match-is-found-with-vanilla-javascript/
+     */
+    Utils.nextUntil = function (el, selector, filter) { //TODO: captialization of nextUntil?
```
Is this capitalized correctly?

stephaniehobson commented on src/assets/js/protocol/protocol-utils.js:
```diff
@@ -0,0 +1,59 @@
+    }
+
+    /**
+     * wrapAll
```
s/wrapAll/nextUntil

alexgibson commented on src/assets/js/protocol/protocol-details.js:
```diff
@@ -0,0 +1,229 @@
+        var fake;
+        var root;
+        var diff;
+        if (!('open' in el)) {
```
Nit: could add some spacing to this code just to make it a little more readable?

alexgibson commented on src/assets/js/protocol/protocol-details.js:
```diff
@@ -0,0 +1,229 @@
+        var details = doc.getElementById(id);
+        var isClosed = details.getAttribute('aria-hidden');
+
+        if(isClosed == 'true') {
```
Nit: spaces before braces here and below

alexgibson commented on src/assets/js/protocol/protocol-footer.js:
```diff
@@ -0,0 +1,21 @@
+(function() {
+    'use strict';
+
+    var _mqWide = matchMedia('(max-width: 479px)');
```
We should add a basic feature detect for `matchMedia` here, just to stop older browsers throwing errors? I did one for the Menu component [here](https://github.com/mozilla/protocol/blob/master/src/assets/js/protocol/protocol-menu.js#L303-L307)

Doesn't need to be part of this PP, but we should file an issue to discuss where we should keep feature detects, as they could probably be consolidated in a library together.

stephaniehobson replied:
Moved the feature detect here and in the Menu into `protocol-supports.js`.

alexgibson commented on src/assets/js/protocol/protocol-utils.js:
```diff
@@ -0,0 +1,59 @@
+        return siblings;
+    };
+
+    window.Mozilla.Utils = Utils;
```
Oh, here's something we haven't thought through quite yet. It's probably only a problem on bedrock, but we already have a `Mozilla.Utils` namespace there :(

We should maybe file an issue and think about namespacing our protocol JS to something more unique `MZP.`?

For now though, maybe we could use `DOMUtils`?

stephaniehobson replied:
Oh we should just be using a MZP namespace globally.

stephaniehobson replied:
Damn that's going to take some changing. I'll add an issue, we can discuss it on Thursday.

stephaniehobson replied:
I'm using `mzp` everywhere now too.

alexgibson commented on src/assets/sass/protocol/base/elements/_details.scss:
```diff
@@ -0,0 +1,46 @@
+.mzp-c-details {
+    .is-summary {
+        button {
+            all: inherit;
```
Browser support is [sketchy](https://developer.mozilla.org/en-US/docs/Web/CSS/all#Browser_compatibility) here, is this usecase safe enough?

stephaniehobson replied:
Updated to declare specific properties to inherit.

alexgibson commented on src/assets/sass/protocol/components/_menu.scss:
```diff
@@ -70,8 +70,7 @@
         &:before {
-            // base64 encoding of https://github.com/mozilla/protocol-assets/blob/master/icons/ui/expand-black.svg
-            background:  url('data:image/svg+xml;base64,PHN2ZyB3aWR0aD0iMTYiIGhlaWdodD0iMTYiIHhtbG5zPSJodHRwOi8vd3d3LnczLm9yZy8yMDAwL3N2ZyI+PHBhdGggZD0iTTcgN1YxYTEgMSAwIDEgMSAyIDB2Nmg2YTEgMSAwIDEgMSAwIDJIOXY2YTEgMSAwIDEgMS0yIDBWOUgxYTEgMSAwIDEgMSAwLTJoNnoiIGZpbGw9IiMwMDAiIGZpbGwtcnVsZT0iZXZlbm9kZCIvPjwvc3ZnPgo=') top left no-repeat;
+            background:  $url-image-expand-black top left no-repeat;
```
🔥

alexgibson commented on src/patterns/molecules/details/details-custom.hbs:
```diff
@@ -0,0 +1,38 @@
+---
+name: Custom expand and collapse
+description: The functions driving the details and summary component can be called independatnly
```
Typo `independatnly`

alexgibson commented on src/assets/js/protocol/protocol-details.js:
```diff
@@ -0,0 +1,229 @@
+    }
+
+    // init generic class indicating headings should be made into expand/collapse componenet
+    Details.init('.mzp-c-details > h2');
```
It would be good to write some JS tests for this component? This might mean separating out the calls to `init` in a separate file.

alexgibson commented on src/patterns/molecules/details/details-custom.hbs:
```diff
@@ -0,0 +1,38 @@
+<script src="{{@root.baseurl}}/assets/protocol/protocol/js/protocol-utils.js"></script>
+<script src="{{@root.baseurl}}/assets/protocol/protocol/js/protocol-details.js"></script>
+<script>
+  var _mqWide = matchMedia('(max-width: 479px)');
```
This isn't executing, it should be inside an IIFE and also out of global scope

alexgibson commented on src/patterns/organisms/footer/footer-min.hbs:
```diff
@@ -0,0 +1,33 @@
+  A page footer organism, containing common links to compliment primary navigation.
+notes: |
+  - The example above shows an embedded Language Switcher component inside the Footer organism.
+  - Inline JavaScript in this example is for demo purposes only. Use external files in production code.
```
There's no inline JS in this example

alexgibson commented on src/patterns/organisms/footer/footer-full.hbs:
```diff
@@ -0,0 +1,51 @@
+          Content available under a <a rel="license" href="https://www.mozilla.org/foundation/licensing/website-content/">Creative Commons license</a>.
+        </p>
+        <ul class="mzp-c-footer-terms">
+          <li><a rel="nofollow" href="https://www.mozilla.org/privacy/websites/">Website Privacy Notice</a></li>
```
We can remove `rel="nofollow"` on these links (@rraue removed them from the bedrock footer)

alexgibson commented on src/patterns/organisms/footer/includes/footer-section-company.hbs:
```diff
@@ -0,0 +1,10 @@
+    Company
+  </h5>
+  <ul class="mzp-c-footer-list">
+    <li><a href="https://www.mozilla.org/en-US/about/">About</a></li>
```
Nit: remove the hardcoded locales from example links. We would know to remove them in bedrock, but others teams might cut and paste from the examples here.

alexgibson commented on src/assets/js/protocol/protocol-details.js:
```diff
@@ -0,0 +1,229 @@
+     * Init
+     * @param {Object} selector - CSS selector matching "summary" elements
+     */
+    Details.init = function(selector) {
```
Since this code path depends on `Mozilla.Utils.nextUntil`, should we add a check that it is defined so it fails gracefully if not?

stephaniehobson replied:
Added a check 👍

alexgibson commented on src/assets/js/protocol/protocol-footer.js:
```diff
@@ -0,0 +1,21 @@
+    var _mqWide = matchMedia('(max-width: 479px)');
+
+    if (_mqWide.matches) {
+        window.Mozilla.Details.init('.mzp-c-footer-sections .mzp-c-footer-heading');
```
Should we also add a check that `Mozilla.Details` is defined before initializing?

stephaniehobson replied:
Added a check 👍

alexgibson commented on src/assets/js/protocol/protocol-details.js:
```diff
@@ -0,0 +1,229 @@
+     * open
+     * @param {String} id - id of the container to open
+     */
+    Details.open = function(id) {
```
Would it be worth adding some callbacks for open and close, so we can add GA dataLayer events in bedrock?

stephaniehobson replied:
Added the ability to do callbacks.

alexgibson commented on src/assets/js/protocol/protocol-details.js:
```diff
@@ -0,0 +1,229 @@
+        // add listener
+        control.addEventListener('click', Details.handleControlActivation, false);
+        // copy the summary's contents into the control
+        summary.childNodes.forEach(function(child) {
```
This is throwing an error in IE

```
SCRIPT5007: Unable to get property 'childNodes' of undefined or null reference 
```

stephaniehobson replied:
Fixed! 🔨

alexgibson commented:
Sorry for the sporatic comments here, github has been playing up saying "Start Review" for every comment :/

alexgibson commented:
Oh, I also noticed that the expand icons aren’t animating on the demo page.

stephaniehobson commented:
No tests yet but everything else should be addressed.

stephaniehobson commented:
Now with tests! 🎉 

Ready for re-review. Probably best to re-review the functionality too since there's been substantial changes.

stephaniehobson commented:
Test failure looks like a dependency problem and not related to my changes?

alexgibson commented:
> Test failure looks like a dependency problem and not related to my changes?

I filed https://github.com/mozilla/bedrock/issues/6481 for this and will open a PR 👍

alexgibson commented:
When testing in IE (both 10 & 11), the very bottom footer accordion on the demo page does not work. There don't seem to be any immediate errors thrown, but I'm guessing it might be because it follows an empty footer? Seems to work fine in Firefox/Chrome.

<img width="380" alt="screen shot 2018-11-16 at 12 39 13" src="https://user-images.githubusercontent.com/400117/48621897-f5a1ed00-e99c-11e8-9587-8813a4f1907b.png">

alexgibson commented:
The small legal links right in the botom of the footer don't have any obvious hover/focus states:

![screen shot 2018-11-16 at 12 43 18](https://user-images.githubusercontent.com/400117/48621967-36016b00-e99d-11e8-9ba2-19991244d20b.png)

stephaniehobson commented on src/patterns/molecules/details/details-custom.hbs:
```diff
@@ -0,0 +1,45 @@
+---
+
+<style>
+.demo-section .is-closed[aria-hidden=true] { display: none; }
```
I'm not sure what you mean here.

alexgibson replied:
I was questioning why this needs to be an inline style, which is something we don't generally encourage. I guess I don't yet fully understand why the custom option doesn't have any basic CSS once it's interactive to cover hiding/showing. Maybe just add a note that the inline styles are for demo purposes only, and should be included in the website's regular stylesheet?

stephaniehobson replied:
I understand now thanks 😃 

I avoided defining `.is-closed[aria-hidden=true] { display: none; }` to the global styles because of the possibility of it interfering with a transition. But it's not obvious the demo is working without it. 

The demo seems to not be communicating what I want it to, I'll think about it for a day and see if I can improve it 📈

stephaniehobson commented on src/patterns/molecules/details/details-custom.hbs:
```diff
@@ -0,0 +1,45 @@
+</style>
+
+<section class="demo-section">
+  <h5 class="demo-heading">Heading</h5>
```
It should be interactive if the window is 767 or smaller. The idea was to demo custom initialization and destroy. Do you have any ideas for a better demo?

alexgibson replied:
Ah I see! This totally escaped me when I was reading the page, and I should have paid closer attention to the code in the example itself. Still, it could perhaps help if the docs described what the example itself is set up to do? I probably wouldn't have been tripped up in assuming something was broken with an accompanying description?

stephaniehobson commented on tests/unit/protocol-details.js:
```diff
@@ -0,0 +1,252 @@
+        });
+
+        it('does not initalize if the browser does not support it', function() {
+            Mzp.Details.isSupported = false;
```
The function is in a closure so it only run the first time the file loads. It doesn't run with each test so threre's nothing to spy on.

alexgibson replied:
Ok, in that case we should at least make sure that the value you're overriding gets reset back to the browser's initial value after each test, so we don't need to rely on things like the order of the tests in order to get everything to pass. Alternatively, you could just make this a simple function and not a closure. I don't think there would be much of a perf hit here since it's only being called once on `init()`.

stephaniehobson replied:
OMG THANK YOU. I spent 5 hours on this on Friday.

stephaniehobson replied:
Okay, I took this out of the closure and update the tests to spy on it.

stephaniehobson commented on src/assets/js/protocol/protocol-details.js:
```diff
@@ -0,0 +1,244 @@
+        var details = doc.getElementById(id);
+        var isClosed = details.getAttribute('aria-hidden');
+
+        if(isClosed == 'true') {
```
I wonder why ESLint didn't pick this up....?

stephaniehobson commented on src/assets/js/protocol/protocol-footer.js:
```diff
@@ -0,0 +1,36 @@
+    'use strict';
+
+    // check we have global variable
+    if (typeof window.Mzp !== 'undefined') {
```
Are you suggesting I change it from `window.Mzp` to `Mzp` or suggesting I define the object if it doesn't exist?

alexgibson replied:
Oh sorry, I see how this comment was confusing. I didn't pay enough attention to the purpose of this file. I had assumed it was another JS library like the others, but I see it's just initialising the footer itself and not much more. I think this is fine as-is 👍

stephaniehobson commented on src/assets/sass/protocol/components/_footer.scss:
```diff
@@ -12,111 +12,231 @@
+        text-decoration: underline;
+    }
+
+    @media (min-width: 480px) and (max-width: 1055px) {
```
They're based off our Media Queries but I couldn't find tokens for the values outside of media queries. Can I add them? (What makes something a token instead of a SCSS variable?)

alexgibson replied:
Interesting, these values should be in our tokens, but I think you may have just discovered a bug: https://github.com/mozilla/protocol-tokens/blob/master/tokens/content.yml#L21-L30

The values for things like `screen-lg` are mapped to `content` values, whereas they should be mapped to the values here https://github.com/mozilla/protocol-tokens/blob/master/tokens/_aliases.yml#L134-L138

alexgibson replied:
I filed an issue for this here https://github.com/mozilla/protocol-tokens/issues/27

alexgibson replied:
PR to expose correct screen size tokens: https://github.com/mozilla/protocol-tokens/pull/28

stephaniehobson commented on src/assets/sass/protocol/components/_footer.scss:
```diff
@@ -12,111 +12,231 @@
+        margin-bottom: 0;
+        padding: $spacing-md 0;
+        position: relative;
+        @include bidi(((text-align, left, right),));
```
`text-align` usually goes between `position` and `width` in the dictionary 😜 Do you want the includes all grouped at the top?

alexgibson replied:
Yes, I'm just suggesting we follow convention for how we do this elsewhere. I'd rather give up ordering decisions to my text editor, rather than us have to start to maintain these kind of things manually 😜

stephaniehobson commented on src/assets/js/protocol/protocol-details.js:
```diff
@@ -0,0 +1,244 @@
+        summarySiblings = Mzp.Utils.nextUntil(summary, selector);
+
+        // look to see if all children are already in a wrapper we can use
+        if(summarySiblings.length == 1) {
```
Oh no. Fixing this broke the test 😭

alexgibson replied:
TL:DR - changing this line was a red herring for the tests failing (the problem lies elsewhere, see my other comment).

stephaniehobson commented:
@alexgibson I can't reproduce the bottom accordion not working in IE 10 or 11. Want to pull down the updated PR and try again. Maybe one of the other review changes fixed the problem.

Also, I removed the underline from the small links in the footer on hover/focus and added an underline to the social media icons on hover/focus

stephaniehobson commented:
Okay, I *think* everything is done except:
- a more use friendly demo
- using the new tokens for the media queries (we need to publish the new tokens to npm first)

alexgibson commented:
@stephaniehobson I published a new version of protcol-tokens to npm, so you should be ready to update and use the new version in this PR. If you could also merge https://github.com/mozilla/protocol/pull/271 and then rebase, we can make sure the unit tests added here are passing in Travis 👍

stephaniehobson commented:
Updated to use the tokens.

Updated to make the demo uglier. Is it easier to understand?

alexgibson commented on package.json:
```diff
@@ -25,7 +25,7 @@
     "@cloudfour/hbs-helpers": "^0.9.0",
     "@mozilla-protocol/assets": "^1.0.1",
-    "@mozilla-protocol/tokens": "^2.0.0",
+    "@mozilla-protocol/tokens": "^2.1.0",
```
This is a change that will require refactoring, so it should probably be a major version bump. 

Still, I probably wouldn't change the version number in this PR unless you plan on publishing to NPM straight away. There may be other changes we can get into a major release?

If you do want to publish right away, you need to follow the [instructions in the docs](https://github.com/mozilla/protocol/tree/master/docs#publishing-to-npm) to do it fully, as changing this alone isn't quite enough. You'll also need to make sure you're set up as a team contributor on NPM.

alexgibson commented on CHANGELOG.md:
```diff
@@ -1,8 +1,14 @@
 ## Bug Fixes
 
 * **js:** Add missing del dependency in build (#271).
+* Update tokens
```
Add a link to the tokens issue so folks can work out what changed in the update?

alexgibson commented on src/patterns/organisms/footer/footer-full.hbs:
```diff
@@ -0,0 +1,53 @@
+  </div>
+</footer>
+<script src="{{@root.baseurl}}/assets/protocol/protocol/js/protocol-utils.js"></script>
+<script src="{{@root.baseurl}}/assets/protocol/protocol/js/protocol-supports.js"></script>
```
Should `protocol-supports.js` and `protocol-utils.js` be included in the base template for the docs site? I notice the navigation now has a dependency on the feature detect library, but it's not included in the navigation demo page, so it's falling back to the CSS-only interactions.

stephaniehobson replied:
Added to the site template.

craigcook commented on src/patterns/organisms/footer/footer-min.hbs:
```diff
@@ -0,0 +1,31 @@
+name: Flexable footer component
+order: 1
+description: |
+  A page footer organism, containing common links to compliment primary navigation.
```
sp: "complement" (technically a homonym and not a misspelling... complement with an e means they go together to complete a perfect whole, compliment with an i means the footer told the main nav it looks really nice today.)

craigcook commented on src/patterns/molecules/details/details-component.hbs:
```diff
@@ -0,0 +1,20 @@
+order: 2
+notes: |
+    - Assumes all content following a heading is intended to be hidden.
+    - Only collapses headings highest in the heirarchy.
```
sp: "hierarchy"

craigcook commented on src/patterns/molecules/details/details-custom.hbs:
```diff
@@ -0,0 +1,75 @@
+---
+name: Custom expand and collapse
+description: The functions driving the details and summary component can be called independantly.
```
sp: "independently"

craigcook commented:
Actually I do see a bug, I'm getting `NotFoundError: Node was not found` locally in `protocol-supports.js` line 49, but I don't get that error on demo1, so possibly a regression or something with my local dev setup?

stephaniehobson commented:
What page are you getting the error on?

stephaniehobson commented:
Found and fixed the JS error.
